File: src/main/java/ca/yyx/hu/UsbAttachedActivity.java
package ca.yyx.hu;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbManager;
import android.os.Bundle;
import android.widget.Toast;

import ca.yyx.hu.aap.AapActivity;
import ca.yyx.hu.aap.AapService;
import ca.yyx.hu.usb.UsbDeviceCompat;
import ca.yyx.hu.usb.UsbModeSwitch;
import ca.yyx.hu.utils.IntentUtils;
import ca.yyx.hu.utils.Settings;
import ca.yyx.hu.utils.Utils;

/**
 * @author algavris
 * @date 30/05/2016.
 */
public class UsbAttachedActivity extends Activity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Utils.logd("USB Intent: " + getIntent());

        UsbDevice device = IntentUtils.getDevice(getIntent());
        if (device == null) {
            Utils.loge("No USB device");
            finish();
            return;
        }

        if (UsbDeviceCompat.isInAccessoryMode(device)) {
            Utils.loge("Usb in accessory mode");
            startService(AapService.createIntent(device, this));
            finish();
            return;
        }

        UsbDeviceCompat deviceCompat = new UsbDeviceCompat(device);
        Settings settings = new Settings(this);
        if (!settings.isConnectingDevice(deviceCompat)) {
            Utils.logd("Skipping device " + deviceCompat.getUniqueName());
            finish();
            return;
        }

        UsbManager usbManager = (UsbManager) getSystemService(Context.USB_SERVICE);
        UsbModeSwitch usbMode = new UsbModeSwitch(usbManager);
        Utils.logd("Switching USB device to accessory mode " + deviceCompat.getUniqueName());
        Toast.makeText(this, "Switching USB device to accessory mode " + deviceCompat.getUniqueName(), Toast.LENGTH_SHORT).show();
        if (usbMode.switchMode(device)) {
            Toast.makeText(this, "Success", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, "Failed", Toast.LENGTH_SHORT).show();
        }

        finish();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        UsbDevice device = IntentUtils.getDevice(getIntent());
        if (device == null) {
            Utils.loge("No USB device");
            finish();
            return;
        }

        Utils.logd(UsbDeviceCompat.getUniqueName(device));

        if (UsbDeviceCompat.isInAccessoryMode(device)) {
            Utils.loge("Usb in accessory mode");
            startService(AapService.createIntent(device, this));
        }

        finish();
    }
}

File: src/main/java/ca/yyx/hu/app/UsbAttachedActivity.kt
package ca.yyx.hu.app

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.hardware.usb.UsbManager
import android.os.Bundle
import android.widget.Toast

import ca.yyx.hu.App
import ca.yyx.hu.aap.AapService
import ca.yyx.hu.connection.UsbDeviceCompat
import ca.yyx.hu.connection.UsbModeSwitch
import ca.yyx.hu.utils.AppLog
import ca.yyx.hu.utils.LocalIntent
import ca.yyx.hu.utils.Settings

/**
 * @author algavris
 * *
 * @date 30/05/2016.
 */
class UsbAttachedActivity : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        AppLog.i("USB Intent: " + intent)

        val device = LocalIntent.extractDevice(intent)
        if (device == null) {
            AppLog.e("No USB device")
            finish()
            return
        }

        if (App.provide(this).transport.isAlive) {
            AppLog.e("Thread already running")
            finish()
            return
        }

        if (UsbDeviceCompat.isInAccessoryMode(device)) {
            AppLog.e("Usb in accessory mode")
            startService(AapService.createIntent(device, this))
            finish()
            return
        }

        val deviceCompat = UsbDeviceCompat(device)
        val settings = Settings(this)
        if (!settings.isConnectingDevice(deviceCompat)) {
            AppLog.i("Skipping device " + deviceCompat.uniqueName)
            finish()
            return
        }

        val usbManager = getSystemService(Context.USB_SERVICE) as UsbManager
        val usbMode = UsbModeSwitch(usbManager)
        AppLog.i("Switching USB device to accessory mode " + deviceCompat.uniqueName)
        Toast.makeText(this, "Switching USB device to accessory mode " + deviceCompat.uniqueName, Toast.LENGTH_SHORT).show()
        if (usbMode.switchMode(device)) {
            Toast.makeText(this, "Success", Toast.LENGTH_SHORT).show()
        } else {
            Toast.makeText(this, "Failed", Toast.LENGTH_SHORT).show()
        }

        finish()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        val device = LocalIntent.extractDevice(getIntent())
        if (device == null) {
            AppLog.e("No USB device")
            finish()
            return
        }

        AppLog.i(UsbDeviceCompat.getUniqueName(device))

        if (!App.provide(this).transport.isAlive) {
            if (UsbDeviceCompat.isInAccessoryMode(device)) {
                AppLog.e("Usb in accessory mode")
                startService(AapService.createIntent(device, this))
            }
        } else {
            AppLog.e("Thread already running")
        }

        finish()
    }
}
